README
NAME
    Lego::Part - Lego part object.

SYNOPSIS
     use Lego::Part;

     my $obj = Lego::Part->new;
     my $color = $obj->color($color);
     my $design_id = $obj->design_id($design_id);
     my $element_id = $obj->element_id($element_id);

METHODS
    *       "new()"

             Constructor.
             Returns object.

    *       "color([$color])"

             Get or set color.
             Returns string with color.

    *       "design_id([$design_id])"

             Get or set lego design id.
             Returns string with design ID.

    *       "element_id([$element_id])"

             Get or set lego element id.
             Returns string with element ID.

EXAMPLE1
     use strict;
     use warnings;

     use Lego::Part;

     # Object.
     my $part = Lego::Part->new(
             'color' => 'red',
             'design_id' => '3002',
     );

     # Print color and design ID.
     print 'Color: '.$part->color."\n";
     print 'Design ID: '.$part->design_id."\n";

     # Output:
     # Color: red
     # Design ID: 3002

EXAMPLE2
     use strict;
     use warnings;

     use Lego::Part;

     # Object.
     my $part = Lego::Part->new(
             'element_id' => '300221',
     );

     # Print color and design ID.
     print 'Element ID: '.$part->element_id."\n";

     # Output:
     # Element ID: 300221

DEPENDENCIES
    Class::Utils, Error::Pure.

SEE ALSO
    Task::Lego
        Install the Lego modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Lego-Part>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2013-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.05



Changes
0.05

0.04 2023-11-24T22:24:12+01:00
 - Add tests.
 - Fix Makefile.PL fails with no '.' in @INC.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Move bugtracker to github.
 - Rename example files to better names.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.03 2015-09-23T23:30:47+02:00
 - Add tests for Lego::Part class.
 - Move author tests to xt/ directory.

0.02 2015-08-24T11:31:45+02:00
 - Add SEE ALSO section in doc.
 - Add minimal version of Perl.
 - Add missing English dependency.
 - Update Module::Install to 1.16 version.

0.01 2015-03-04T00:48:41+01:00
 - First version.


Part.pm
package Lego::Part;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);

our $VERSION = 0.05;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Color.
	$self->{'color'} = undef;

	# Lego design id.
	$self->{'design_id'} = undef;

	# Lego element id.
	$self->{'element_id'} = undef;

	# Process parameters.
	set_params($self, @params);

	# Check design id or element id.
	if (! defined $self->{'element_id'}
		&& ! defined $self->{'design_id'}) {

		err "Parameter 'element_id' or 'design_id' is required.";
	}

	# Object.
	return $self;
}

# Get or set color.
sub color {
	my ($self, $color) = @_;
	if ($color) {
		$self->{'color'} = $color;
	}
	return $self->{'color'};
}

# Get or set lego design id.
sub design_id {
	my ($self, $design_id) = @_;
	if ($design_id) {
		$self->{'design_id'} = $design_id;
	}
	return $self->{'design_id'};
}

# Get or set lego element id.
sub element_id {
	my ($self, $element_id) = @_;
	if ($element_id) {
		$self->{'element_id'} = $element_id;
	}
	return $self->{'element_id'};
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part - Lego part object.

=head1 SYNOPSIS

 use Lego::Part;

 my $obj = Lego::Part->new;
 my $color = $obj->color($color);
 my $design_id = $obj->design_id($design_id);
 my $element_id = $obj->element_id($element_id);

=head1 METHODS

=over 8

=item * C<new()>

 Constructor.
 Returns object.

=item * C<color([$color])>

 Get or set color.
 Returns string with color.

=item * C<design_id([$design_id])>

 Get or set lego design id.
 Returns string with design ID.

=item * C<element_id([$element_id])>

 Get or set lego element id.
 Returns string with element ID.

=back

=head1 EXAMPLE1

=for comment filename=create_part_and_print1.pl

 use strict;
 use warnings;

 use Lego::Part;

 # Object.
 my $part = Lego::Part->new(
         'color' => 'red',
         'design_id' => '3002',
 );

 # Print color and design ID.
 print 'Color: '.$part->color."\n";
 print 'Design ID: '.$part->design_id."\n";

 # Output:
 # Color: red
 # Design ID: 3002

=head1 EXAMPLE2

=for comment filename=create_part_and_print2.pl

 use strict;
 use warnings;

 use Lego::Part;

 # Object.
 my $part = Lego::Part->new(
         'element_id' => '300221',
 );

 # Print color and design ID.
 print 'Element ID: '.$part->element_id."\n";

 # Output:
 # Element ID: 300221

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>.

=head1 SEE ALSO

=over

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2013-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.05

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Lego part object.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Lego-Part';
readme_from 'Part.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.03;
requires 'English' => 0;
requires 'Error::Pure' => 0.22;
requires 'Scalar::Util' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Lego-Part/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Lego-Part';
resources 'repository' => 'git://github.com/michal-josef-spacek/Lego-Part';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.05';

# Run.
WriteAll();


Part/Transfer.pm
package Lego::Part::Transfer;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use Scalar::Util qw(blessed);

our $VERSION = 0.05;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Process parameters.
	set_params($self, @params);

	# Object.
	return $self;
}

# Convert element to design.
sub element2design {
	my ($self, $part) = @_;
	$self->_check_part($part);
	err "This is abstract class. element2design() method not implemented.";
	return;
}

# Convert design to element.
sub design2element {
	my ($self, $part) = @_;
	$self->_check_part($part);
	err "This is abstract class. design2element() method not implemented.";
	return;
}

# Check part class.
sub _check_part {
	my ($self, $part) = @_;
	if (! blessed($part) || ! $part->isa('Lego::Part')) {
		err "Part must be Lego::Part object.";
	}
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Transfer - Lego part transfer abstract class.

=head1 SYNOPSIS

 use Lego::Part::Transfer;

 my $obj = Lego::Part::Transfer->new;
 $obj->element2design($part);
 $obj->design2element($part);

=head1 METHODS

=over 8

=item * C<new()>

 Constructor.
 Returns object.

=item * C<element2design($part)>

 Convert element to design.
 Returns undef.

=item * C<design2element($part)>

 Convert design to element.
 Returns undef.

=back

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2013-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.05

=cut


Part/Action.pm
package Lego::Part::Action;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English;
use Error::Pure qw(err);
use Scalar::Util qw(blessed);

our $VERSION = 0.05;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Process parameters.
	set_params($self, @params);

	# Object.
	return $self;
}

# Load design id to Lego::Part object.
sub load_design_id {
	my ($self, $part_transfer_class, $part) = @_;
	$self->_check_part_transfer_class($part_transfer_class);
	eval {
		$part_transfer_class->element2design($part);
	};
	if ($EVAL_ERROR) {
		err 'Cannot load design ID.',
			'Error', $EVAL_ERROR;
	}
	return;
}

# Load element id to Lego::Part object.
sub load_element_id {
	my ($self, $part_transfer_class, $part) = @_;
	$self->_check_part_transfer_class($part_transfer_class);
	eval {
		$part_transfer_class->design2element($part);
	};
	if ($EVAL_ERROR) {
		err 'Cannot load element ID.',
			'Error', $EVAL_ERROR;
	}
	return;
}

# Check transfer class.
sub _check_part_transfer_class {
	my ($self, $part_transfer_class) = @_;
	if (! blessed($part_transfer_class)
		|| ! $part_transfer_class->isa('Lego::Part::Transfer')) {

		err "Bad transfer class. Must be 'Lego::Part::Transfer' ".
			'class.';
	}
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Action - Lego part action object.

=head1 SYNOPSIS

 use Lego::Part::Action;

 my $obj = Lego::Part::Action->new;
 $obj->load_design_id($part_transfer_class, $part);
 $obj->load_element_id($part_transfer_class, $part);

=head1 METHODS

=over 8

=item * C<new()>

 Constructor.
 Returns object.

=item * C<load_design_id($part_transfer_class, $part)>

 Load design id to Lego::Part object.
 Returns undef.

=item * C<load_element_id($part_transfer_class, $part)>

 Load element id to Lego::Part object.
 Returns undef.

=back

=head1 ERRORS

 load_design_id():
         Bad transfer class. Must be 'Lego::Part::Transfer' class.
         Cannot load design ID.
                 Error: %s
 load_element_id():
         Bad transfer class. Must be 'Lego::Part::Transfer' class.
         Cannot load element ID.
                 Error: %s

=head1 EXAMPLE

=for comment filename=design_to_element_and_print.pl

 package Lego::Part::Transfer::Example;

 use base qw(Lego::Part::Transfer);
 use strict;
 use warnings;

 # Convert design to element.
 sub design2element {
         my ($self, $part) = @_;
         $self->_check_part($part);
         if ($part->color eq 'red' && $part->design_id eq '3002') {
                 $part->element_id('300221');
         }
         return;
 }

 package main;

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Action;

 # Lego part.
 my $part = Lego::Part->new(
         'color' => 'red',
         'design_id' => '3002',
 );

 #  Lego part action.
 my $act = Lego::Part::Action->new;

 # Transfer class.
 my $trans = Lego::Part::Transfer::Example->new;

 # Load element id.
 $act->load_element_id($trans, $part);

 # Print color and design ID.
 print 'Color: '.$part->color."\n";
 print 'Design ID: '.$part->design_id."\n";
 print 'Element ID: '.$part->element_id."\n";

 # Output:
 # Color: red
 # Design ID: 3002
 # Element ID: 300221

=head1 DEPENDENCIES

L<Class::Utils>,
L<English>,
L<Error::Pure>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2013-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.05

=cut
